helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -241,21 +241,64 @@ private void refreshClusterConfig(final HelixDataAccessor accessor,
     if (_propertyDataChangedMap.get(HelixConstants.ChangeType.CLUSTER_CONFIG).getAndSet(false)) {
       _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
       refreshedType.add(HelixConstants.ChangeType.CLUSTER_CONFIG);
+      // TODO: This is a temp function to clean up incompatible batched disabled instances format.
+      // Remove in later version.
+      if (_clusterConfig!=null && needCleanUpBatchedDisabledInstance(_clusterConfig.getRecord())
+          && cleanBatchDisableMapField(accessor)) {
+        LogUtil.logInfo(logger, getClusterEventId(), String
+            .format("Clean ClusterConfig mapField for cluster %s, pipeline %s", _clusterName,
+                getPipelineName()));
+      }
       refreshAbnormalStateResolverMap(_clusterConfig);
     } else {
-      LogUtil.logInfo(logger, getClusterEventId(), String.format(
-          "No ClusterConfig change for cluster %s, pipeline %s", _clusterName, getPipelineName()));
+      LogUtil.logDebug(logger, getClusterEventId(), String
+          .format("No ClusterConfig change for cluster %s, pipeline %s", _clusterName,
+              getPipelineName()));
     }
   }
 
+  // TODO: This function is used to clean up batched disabled instances for
+  // "DISABLED_INSTANCES" introduced in 1.0.3.0. This temp change should be reverted after 1.0.5.0 \
+  // or later version.
+  private boolean cleanBatchDisableMapField(final HelixDataAccessor accessor) {
+    boolean successful =
+        accessor.updateProperty(accessor.keyBuilder().clusterConfig(), new DataUpdater<ZNRecord>() {
+          @Override
+          public ZNRecord update(ZNRecord currentData) {
+            if (currentData == null) {
+              throw new HelixException(
+                  "Cluster: " + _clusterConfig.getClusterName() + ": cluster config is null");
+            }
+            ZNRecord newRecord = new ZNRecord(currentData);
+            String batchDisabledInstanceMapFieldKey =
+                ClusterConfig.ClusterConfigProperty.DISABLED_INSTANCES.name();
+            if (needCleanUpBatchedDisabledInstance(currentData)) {
```

Comment:
you are calling needCleanUpBatchedDisabledInstance() twice.
first on line:246 and here too..  any particular reason

Reply:
It is possible that there is a race condition that 2 pipeline both check clusterConfig that need update. By adding another check in updater, we reduce the race window and updater also garente that only one will write back.

Reply:
in my mind, first condition doesn't help you. to avoid race, you do it here, should be sufficient.

Reply:
I think the first if helps. If that first condition did not satisfy, we will safe a read of this updater.

Reply:
in my mind, doing it unconditionally once is sufficient. reading/update as one step (method) is easier to reason through, but i am ok with this 2-step approach